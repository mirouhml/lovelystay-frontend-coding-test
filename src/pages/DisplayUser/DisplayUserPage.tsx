import React from 'react';
import UserInfo from '../../components/UserInfo/UserInfo';

const DisplayUserPage = () => {
  return (
    <div>
      <UserInfo />
    </div>
  );
};

export default DisplayUserPage;
