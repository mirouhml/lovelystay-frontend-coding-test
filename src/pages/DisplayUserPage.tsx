import React, { ReactElement } from 'react';
import UserInfoCard from '../components/UserInfoCard';
import RepositoriesList from '../components/RepositoriesList';
import { IoIosArrowBack } from 'react-icons/io';
import { useNavigate } from 'react-router-dom';

const DisplayUserPage = (): ReactElement => {
  const navigate = useNavigate();

  const goBack = () => {
    navigate('..');
  };

  return (
    <div className='display-user-page'>
      <h1>User Result</h1>
      <button
        className='back-button'
        onClick={() => goBack()}
        aria-label='back'
      >
        <IoIosArrowBack size={20} />
      </button>
      <UserInfoCard />
      <RepositoriesList />
    </div>
  );
};

export default DisplayUserPage;
