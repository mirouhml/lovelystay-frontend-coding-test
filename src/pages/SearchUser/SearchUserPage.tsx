import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';

const SearchUserPage = () => {
  const [username, setUsername] = useState('');
  const navigate = useNavigate();
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setUsername(e.target.value);
  };

  const onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    navigate(`/${username}`);
  };

  return (
    <div className='search-user-page'>
      <h1>Search for a GitHub user</h1>
      <form onSubmit={onSubmit} className='form-container'>
        <input
          className='search-input'
          placeholder='Username'
          onChange={onChange}
          required
        />
        <button>Search</button>
      </form>
    </div>
  );
};

export default SearchUserPage;
