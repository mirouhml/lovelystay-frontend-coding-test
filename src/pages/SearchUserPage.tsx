import React, { ReactElement, useState } from 'react';
import { useNavigate } from 'react-router-dom';

const SearchUserPage = (): ReactElement => {
  // state to store the inputted username
  const [username, setUsername] = useState('');
  // hook to navigate to a different page
  const navigate = useNavigate();
  // event handler for input change, updates the username state
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setUsername(e.target.value);
  };
  // event handler for form submission, navigates to a different page with the inputted username as part of the URL
  const onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    navigate(`/${username}`);
  };

  return (
    <div className='search-user-page'>
      {/* Page header */}
      <h1>Search for a GitHub user</h1>
      {/* Form with input field and submit button */}
      <form onSubmit={onSubmit} className='form-container'>
        <input
          className='search-input'
          placeholder='Username'
          onChange={onChange}
          required
        />
        <button>Search</button>
      </form>
    </div>
  );
};

export default SearchUserPage;
