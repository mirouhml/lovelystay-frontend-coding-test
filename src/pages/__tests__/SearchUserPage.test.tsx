import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { RootState } from '../../app/store';
import renderer from 'react-test-renderer';
import SearchUserPage from '../SearchUserPage';
import DisplayUserPage from '../DisplayUserPage';
import store from '../../app/store';

const mockStore = configureStore([thunk]);

describe('SearchUserPage', () => {
  it('should match the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <SearchUserPage />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have a search input', () => {
    render(
      <MemoryRouter>
        <SearchUserPage />
      </MemoryRouter>
    );
    expect(screen.getByPlaceholderText('Username')).toBeInTheDocument();
  });

  it('should have a search button', () => {
    render(
      <MemoryRouter>
        <SearchUserPage />
      </MemoryRouter>
    );
    expect(screen.getByText('Search')).toBeInTheDocument();
  });

  it('should lead to the user page when the search button is clicked', async () => {
    render(
      <Provider
        store={mockStore({
          user: {
            user: {
              id: '1',
              login: 'johndoe',
              name: 'John Doe',
              avatar_url: 'avatar-url',
              public_repos: 0,
              number_of_repo_pages: 0,
              url: 'github-url',
            },
            repositories: [],
            status: {
              user: 'success',
              repos: 'idle',
            },
            error: undefined,
          },
        } as RootState)}
      >
        <MemoryRouter initialEntries={['/']}>
          <Routes>
            <Route path='/' element={<SearchUserPage />} />
            <Route path='/:username' element={<DisplayUserPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    fireEvent.change(screen.getByPlaceholderText('Username'), {
      target: { value: 'johndoe' },
    });
    fireEvent.click(screen.getByText('Search'));
    expect(await screen.findByText('John Doe')).toBeInTheDocument();
  });
});
