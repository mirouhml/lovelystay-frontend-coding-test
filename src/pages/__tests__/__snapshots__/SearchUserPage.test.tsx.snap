// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchUserPage should match the snapshot 1`] = `
<div
  className="search-user-page"
>
  <h1>
    Search for a GitHub user
  </h1>
  <form
    className="form-container"
    onSubmit={[Function]}
  >
    <input
      className="search-input"
      onChange={[Function]}
      placeholder="Username"
      required={true}
    />
    <button>
      Search
    </button>
  </form>
</div>
`;
